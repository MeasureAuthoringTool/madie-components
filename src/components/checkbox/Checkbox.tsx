import React from "react";
import _ from "lodash";
import {
  DisabledDescription,
  DisabledLabel,
  EnabledDescription,
  EnabledLabel,
  Input,
  InputWrapper,
  LabelWrapper,
  OuterWrapper,
} from "./CheckboxStyles";

export interface CheckboxProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  children?: never;
  label: string;
  description?: string;
  disabled?: boolean;
  id?: string;
  checked?: boolean;
  handleChange?: (value: boolean) => void;
  "data-testid"?: string;
}

export default function Checkbox(args: CheckboxProps) {
  const { id, label, description, checked, handleChange, disabled, ...props } =
    args;

  const inputId = id || `checkbox-${_.uniqueId()}`;
  const labelId = `${inputId}-label`;
  const descriptionId = description ? `${inputId}-description` : undefined;

  function changeHandler(e: React.ChangeEvent<HTMLInputElement>) {
    if (handleChange) {
      handleChange(!!e.target.checked);
    }
  }

  const Label = disabled ? DisabledLabel : EnabledLabel;
  const Description = disabled ? DisabledDescription : EnabledDescription;

  return (
    <OuterWrapper>
      <InputWrapper>
        <Input
          id={inputId}
          aria-describedby={descriptionId}
          type="checkbox"
          checked={!!checked}
          onChange={changeHandler}
          disabled={disabled}
          {...props}
        />
      </InputWrapper>
      <LabelWrapper>
        <Label id={labelId} htmlFor={inputId}>
          {label}
        </Label>
        {description ? (
          <Description id={descriptionId}>{description}</Description>
        ) : undefined}
      </LabelWrapper>
    </OuterWrapper>
  );
}
