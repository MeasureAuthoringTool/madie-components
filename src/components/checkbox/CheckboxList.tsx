import React from "react";
import { Fieldset, Legend } from "./CheckboxStyles";
import Checkbox, { CheckboxProps } from "./Checkbox";

export interface CheckboxListProps {
  children?: never;
  legend?: string;
  checkboxes: Array<CheckboxProps>;
}

export default function CheckboxList(args: CheckboxListProps) {
  const { legend, checkboxes, ...props } = args;
  return (
    <Fieldset>
      {legend ? <Legend>{legend}</Legend> : undefined}
      {checkboxes.map((checkbox, index) => (
        <Checkbox key={index} {...checkbox} />
      ))}
    </Fieldset>
  );
}
