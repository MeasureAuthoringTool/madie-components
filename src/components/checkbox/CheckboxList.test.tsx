import * as React from "react";
import { render, fireEvent } from "@testing-library/react";
import CheckboxList from "./CheckboxList";
import { CheckboxProps } from "./Checkbox";

const legendText = "Legendary legend";
const checkbox1: CheckboxProps = {
  label: "Checkbox One",
  ["data-testid"]: "checkbox1",
  checked: true,
};

const checkbox2: CheckboxProps = {
  label: "Checkbox Two",
  ["data-testid"]: "checkbox2",
  checked: false,
};

describe("CheckboxList component", () => {
  it("should match the snapshot", () => {
    const { container: container1 } = render(
      <CheckboxList legend={legendText} checkboxes={[checkbox1, checkbox2]} />
    );
    expect(container1).toMatchSnapshot();
    const { container: container2 } = render(
      <CheckboxList checkboxes={[checkbox1, checkbox2]} />
    );
    expect(container2).toMatchSnapshot();
    const { container: container3 } = render(<CheckboxList checkboxes={[]} />);
    expect(container3).toMatchSnapshot();
  });

  it("should render a CheckboxList with a legend", () => {
    const { getByText, getByTestId } = render(
      <CheckboxList legend={legendText} checkboxes={[checkbox1, checkbox2]} />
    );

    const legend = getByText(legendText) as HTMLLegendElement;
    expect(legend).toBeInstanceOf(HTMLLegendElement);
    expect(legend).toBeInTheDocument();

    const input1 = getByTestId("checkbox1") as HTMLInputElement;
    expect(input1).toBeInstanceOf(HTMLInputElement);
    expect(input1).toBeInTheDocument();
    expect(input1.checked).toBeTruthy();

    const label1 = getByText("Checkbox One") as HTMLLabelElement;
    expect(label1).toBeInstanceOf(HTMLLabelElement);
    expect(label1).toBeInTheDocument();

    const input2 = getByTestId("checkbox2") as HTMLInputElement;
    expect(input2).toBeInstanceOf(HTMLInputElement);
    expect(input2).toBeInTheDocument();
    expect(input2.checked).toBeFalsy();

    const label2 = getByText("Checkbox Two") as HTMLLabelElement;
    expect(label2).toBeInstanceOf(HTMLLabelElement);
    expect(label2).toBeInTheDocument();
  });

  it("should render a CheckboxList with no legend", () => {
    const { container } = render(
      <CheckboxList checkboxes={[checkbox1, checkbox2]} />
    );
    expect(container.childNodes.length).toBe(1);
    const fieldset = container.childNodes[0];
    expect(fieldset).toBeInstanceOf(HTMLFieldSetElement);
    expect(fieldset.childNodes.length).toBe(2); // one for each checkbox
  });
});
