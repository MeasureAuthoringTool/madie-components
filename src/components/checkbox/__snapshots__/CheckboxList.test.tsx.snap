// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxList component should match the snapshot 1`] = `
<div>
  <fieldset
    class="CheckboxStyles__Fieldset-sc-aqujzc-11 fkJecI"
  >
    <legend
      class="CheckboxStyles__Legend-sc-aqujzc-10 dfFTBU"
    >
      Legendary legend
    </legend>
    <div
      class="CheckboxStyles__OuterWrapper-sc-aqujzc-7 WrQcv"
    >
      <div
        class="CheckboxStyles__InputWrapper-sc-aqujzc-8 jzpKC"
      >
        <input
          checked=""
          class="CheckboxStyles__Input-sc-aqujzc-0 chpdNc"
          data-testid="checkbox1"
          id="checkbox-1"
          type="checkbox"
        />
      </div>
      <div
        class="CheckboxStyles__LabelWrapper-sc-aqujzc-9 ejrbED"
      >
        <label
          class="CheckboxStyles__BaseLabel-sc-aqujzc-1 CheckboxStyles__EnabledLabel-sc-aqujzc-2 cWXsqN gnGvtB"
          for="checkbox-1"
          id="checkbox-1-label"
        >
          Checkbox One
        </label>
      </div>
    </div>
    <div
      class="CheckboxStyles__OuterWrapper-sc-aqujzc-7 WrQcv"
    >
      <div
        class="CheckboxStyles__InputWrapper-sc-aqujzc-8 jzpKC"
      >
        <input
          class="CheckboxStyles__Input-sc-aqujzc-0 chpdNc"
          data-testid="checkbox2"
          id="checkbox-2"
          type="checkbox"
        />
      </div>
      <div
        class="CheckboxStyles__LabelWrapper-sc-aqujzc-9 ejrbED"
      >
        <label
          class="CheckboxStyles__BaseLabel-sc-aqujzc-1 CheckboxStyles__EnabledLabel-sc-aqujzc-2 cWXsqN gnGvtB"
          for="checkbox-2"
          id="checkbox-2-label"
        >
          Checkbox Two
        </label>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`CheckboxList component should match the snapshot 2`] = `
<div>
  <fieldset
    class="CheckboxStyles__Fieldset-sc-aqujzc-11 fkJecI"
  >
    <div
      class="CheckboxStyles__OuterWrapper-sc-aqujzc-7 WrQcv"
    >
      <div
        class="CheckboxStyles__InputWrapper-sc-aqujzc-8 jzpKC"
      >
        <input
          checked=""
          class="CheckboxStyles__Input-sc-aqujzc-0 chpdNc"
          data-testid="checkbox1"
          id="checkbox-3"
          type="checkbox"
        />
      </div>
      <div
        class="CheckboxStyles__LabelWrapper-sc-aqujzc-9 ejrbED"
      >
        <label
          class="CheckboxStyles__BaseLabel-sc-aqujzc-1 CheckboxStyles__EnabledLabel-sc-aqujzc-2 cWXsqN gnGvtB"
          for="checkbox-3"
          id="checkbox-3-label"
        >
          Checkbox One
        </label>
      </div>
    </div>
    <div
      class="CheckboxStyles__OuterWrapper-sc-aqujzc-7 WrQcv"
    >
      <div
        class="CheckboxStyles__InputWrapper-sc-aqujzc-8 jzpKC"
      >
        <input
          class="CheckboxStyles__Input-sc-aqujzc-0 chpdNc"
          data-testid="checkbox2"
          id="checkbox-4"
          type="checkbox"
        />
      </div>
      <div
        class="CheckboxStyles__LabelWrapper-sc-aqujzc-9 ejrbED"
      >
        <label
          class="CheckboxStyles__BaseLabel-sc-aqujzc-1 CheckboxStyles__EnabledLabel-sc-aqujzc-2 cWXsqN gnGvtB"
          for="checkbox-4"
          id="checkbox-4-label"
        >
          Checkbox Two
        </label>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`CheckboxList component should match the snapshot 3`] = `
<div>
  <fieldset
    class="CheckboxStyles__Fieldset-sc-aqujzc-11 fkJecI"
  />
</div>
`;
